dnadoh/analyze.py

```diff
@@ -0,0 +1,169 @@
+import argparse
+
+import pandas as pd
+import matplotlib.pyplot as plt
+import plotly.express as px
+import plotly.graph_objects as go
+from plotly.subplots import make_subplots
+
+from plotly.tools import FigureFactory as FF
+from statannot import add_stat_annotation
+from dash import Dash, dcc, html, Input, Output, dash_table
+from scipy import stats
+
+from assemble import assemble
+from plotting_utils import add_pvalue_annotation
+
+BASES = ["A", "C", "G", "T"]
+
+
+def get_unique_locations(df: pd.DataFrame):
+    """Get unique locations with an alternate base"""
+    locations = df["location"].drop_duplicates().dropna().to_list()
+    return locations
+
+
+def _get_pids(df: pd.DataFrame) -> set:
+    pids = set(df["pid"].drop_duplicates().to_list())
+    return pids
+
+
+def plot_boxplot(df: pd.DataFrame, y_variable: str, location: int) -> None:
+    """
+    Creates boxplots of continuous variable 'y' on the y-axis
+    for each reference base.
+    """
+    # subset dataframe to the location
+    subset_df = df[df["location"] == location]
+    # get the reference base
+    reference = set(subset_df["reference"].to_list())
+    # there should only be one
+    assert (
+        len(reference) == 1
+    ), f"There is more than one reference base at position f{location}!"
+    reference = reference.pop()
+
+    # the subset will only contain individuals with an alternate base
+    # to include individuals with the reference base,
+    # we need to figure out who those are and add them back in
+    all_pids = _get_pids(df)
+    subset_pids = _get_pids(subset_df)
+    leftover_pids = all_pids.difference(subset_pids)
+    # build info to concatenate back
```

Comment:
i usually put a blank line before comment lines to make the "paragraphs" easier to see